packages/dapi/test/unit/providers/chainDataProvider.spec.js

```diff
@@ -0,0 +1,280 @@
+  // we should call for 3 blocks (3,4,5) and set 2 missing in the cache
+  it('should use cache when miss something in the tail', async () => {
+    const [first, second, third, fourth, fifth] = fakeHeaders;
+
+    coreAPIMock.getBlockStats.resolves({ height: 1 });
+    coreAPIMock.getBlockHeaders.resolves([third.toString(), fourth.toString(),
+      fifth.toString()]);
+
+    // should use cache and does not hit rpc
+    blockHeadersCache.set(1, first.toString());
+    blockHeadersCache.set(2, second.toString());
+    blockHeadersCache.set(3, third.toString());
+    blockHeadersCache.set(4, undefined);
+    blockHeadersCache.set(5, undefined);
+    cacheSpy.set.resetHistory();
+
+    await chainDataProvider.getBlockHeaders(first.hash, 1, 5);
+
+    expect(coreAPIMock.getBlockHeaders).to.be.calledOnceWithExactly(third.hash, 3);
+
+    expect(cacheSpy.get.callCount).to.be.equal(5);
+    expect(cacheSpy.set.callCount).to.be.equal(3);
+
+    expect(cacheSpy.get.getCall(0).returnValue).to.deep.equal(first.toString());
+    expect(cacheSpy.get.getCall(1).returnValue).to.deep.equal(second.toString());
+    expect(cacheSpy.get.getCall(2).returnValue).to.deep.equal(third.toString());
+    expect(cacheSpy.get.getCall(3).returnValue).to.deep.equal(undefined);
+    expect(cacheSpy.get.getCall(4).returnValue).to.deep.equal(undefined);
+  });
+
+  // the case when we miss something in the middle
+  // f.e we request for 5 blocks, and cache is [1,2,undefined,undefined,5]
+  // should take second block as a start point and request for 4 blocks (to the end)
+  it('should use cache when missing in the middle', async () => {
+    const [first, second, third, fourth, fifth] = fakeHeaders;
+
+    coreAPIMock.getBlockStats.resolves({ height: 1 });
+    coreAPIMock.getBlockHeaders.resolves([second.toString(), third.toString(),
+      fourth.toString(), fifth.toString()]);
+
+    blockHeadersCache.set(1, first.toString());
+    blockHeadersCache.set(2, second.toString());
+    blockHeadersCache.set(3, undefined);
+    blockHeadersCache.set(4, undefined);
+    blockHeadersCache.set(5, fifth.toString());
+    cacheSpy.set.resetHistory();
+
+    await chainDataProvider.getBlockHeaders(first.hash, 1, 5);
+
+    expect(cacheSpy.get.callCount).to.be.equal(5);
+    expect(cacheSpy.set.callCount).to.be.equal(4);
+
+    expect(cacheSpy.get.getCall(0).returnValue).to.deep.equal(first.toString());
+    expect(cacheSpy.get.getCall(1).returnValue).to.deep.equal(second.toString());
+    expect(cacheSpy.get.getCall(2).returnValue).to.deep.equal(undefined);
+    expect(cacheSpy.get.getCall(3).returnValue).to.deep.equal(undefined);
+    expect(cacheSpy.get.getCall(4).returnValue).to.deep.equal(fifth.toString());
+  });
+
+  // the case where we have something in the cache, but the first blocks are not
+  // f.e. we request for 5 blocks, and cache is [undefined,undefined,3,4,5]
+  it('should not use cache when miss something in the beginning', async () => {
```

Comment:
Can you please remind me what was the complexity of using cache for this case?

From what I see, in case we have `[undefined, undefined, 3, 4, 5]`, we still can fetch only the first two items, and serve the last three from the cache.

Am I missing something?